### DISCUSSION TITLE
TickerHandler Usage Question

### DISCUSSION THREAD

**Discussion by RandomEC** [evennia/evennia#2652] [Evennia Questions] [answered]
So, I read the TickerHandler documentation page, and was a bit confused by the "when not to use TickerHandler" at the bottom. I have several features that I am either in the middle of implementing, or am just about to implement, all of which involve some degree of timing. My thoughts on each are below, but I'd appreciate any feedback on the appropriateness of use, or a better solution.

1. Resetting areas - this seems like a prime use of TickerHandler, based on the description in the documentation. (Virtually) every time the ticker goes off, it will/should do something - reset the area.
2. Affects wearing off - as above, as long as a ticker is being created each time an affect goes on, it will only call once - when it is time to remove the affect. It does its thing, and goes.
3. Combat iteration - Though this is a much shorter time, this also seems to fall into the same camp as the above two. Every time it fires, each party in the combat will do its "base" attacks, which are supplemented by whatever skills/spells the combatants want to use.  It is tied to the combat object, so the ticker will only be in use if there is, actually, a combat going on, and only when it is.

These last two were the ones I was less certain of.
4. Mobiles being, well, mobile - Every so often, mobiles that are not sentinels will make a decision about whether they should try to move or not. This won't happen (or at least not successfully) every time, but since a decision does have to be made each time, tickerhandler still seems to be the way to go.
5. Healing/Mana Recovery/Hunger/Thirst - Anyone who is injured will need to heal. As the VAST majority of things that could potentially need to heal are NOT going to need to heal at any given time, this one gave me pause. However, in my game, you heal better depending on your hunger status (at least for players). So, for players at least, I'm not sure how I'd do a "catch up next time you look" model, as hunger is also going to change over time. Additionally, it is certainly important in combat to have a to-the-second idea of what everyone's relative health is. It still feels like the ticker handler is the way to go on this, but maybe I should just have the handler be placed when something takes damage, if it does not already have one? Then the method it calls can remove it when fully healed?

As usual, any thoughts are appreciated. Thanks.

**Comment by mrseeker** [comment]
TickerHandler works only if you have something re-occuring that does not require lots of time, and can get executed "on the spot". A cooldown timer for example does not repeat itself, and is therefore not handy (best to do a timed callback). Same goes for fighting, you don't know when the fight is going to end.

I have a TickerHandler that updates everything in the world like a "world simulation" tick. Unfortunately it's not really optimized and each tick it will lock everything and you need to wait for the whole calculation to end before continuing. Annoying as hell, but it works.

Things that repeat itself every minute/hour/day can be considered useful for the TickerHandler. One-shots and finite tickers are best to be handled with yields and timed callbacks. These offer the same luxury as the TickerHandler, but don't need repeat themselves.
If you answer is no to the following question, you use a TickerHandler: "do I need to delete the timer at a certain point in time?"

**Reply by RandomEC** [reply]
I definitely need to look into delays and timed callbacks more! Thanks!

**Comment by Tegiminis** [comment, marked as answer]
TL;DR: Use TickerHandlers for a simple call you want to repeat at a regular interval, scripts for complicated calls at variable intervals, and `delay()` for a one-off call.

I'm gonna break down the three main ways you would handle an asynchronous call in Evennia (as far as I understand it, anyway), go over what's useful about them, and then talk about where I would use them in your requirements.

1. **TickerHandler**: Performs a function at a periodic, unchanging interval. This is useful for stuff that you always want to happen at a specific interval, and you don't want to change that interval or anything about the callback. For example, I use TickerHandler to convert "temporary XP" to "permanent XP" on players.
2. **Scripts**: Performs the scripted code at the set interval, with a bunch of helper hooks and timer features to allow for more complex logic. Also, counts as a database object, just like normal game objects. Scripts are great for when you want a ticker that is responsive to changes in the game and/or has a variable interval. For example, I use scripts for my mob AI state machine.
3. `utils.delay()`: Performs a function after a set amount of time. This is useful for stuff that will be firing off once and then never again, or which repeats after an evaluation. For example, I use `delay()` to do my effect/buff clean-up on expiry.

Now, for your particular use cases:

1. **Resetting areas** - TickerHandler or Scripts. Functionally similar, the main difference is that for TickerHandler you'll probably want to have an object in the area you attach it to that contains all the information, whereas a Script (as a DB object itself) doesn't need to be attached to anything. I'd probably use a script for this.
2. **Effect expiry** - `delay()`. It's only happening once and it's variable based on the amount of time the effect is on you. I do it this way in my own project.
3. **Combat** - TickerHandler or `delay()`. I personally use `delay()` in my own code because I have some interactions between using other abilities and my attack cooldown that cause interval variation.
4. **Mobiles** - TickerHandler or script. As mentioned above, I use scripts, mainly because my mobs have variable reaction times. I'll probably translate it into a TickerHandler with some `delay()` calls in the future though.
5. **Healing** - TickerHandler. You are making the healing call on a set interval. To make the call dependent on game state, make sure it's a method call on the object (`heal_tick()` or somesuch) which reads the game object to determine its hunger (for the heal strength). Assuming you don't have in-combat health regeneration, I'd say make it a command; otherwise, making a method on the character like `take_damage()` and having it set up the TickerHandler seems like your best bet. Since you can't change the interval with a TickerHandler without removing/reapplying, the "heal strength" here would need to be the amount you heal, rather than the speed.

**Reply by Tegiminis** [reply]
One odd edge case for these that I have is that I use recursive `delay()` calls to perform ticking buffs (ex: poison). I don't use a TickerHandler because I don't instantiate the buff until its call is necessary, and I don't use a script because it's a game object and that seems excessive for my needs.

**Reply by RandomEC** [reply]
Man, one of these days I need to just read all of the functions in utils. There's so much good stuff in there. Ironically, I completely forgot that I already settled on an answer for #1. For what it's worth, I DID go with a timed script, which works nicely.

2. This makes total sense.
3. My preliminary testing on this has me fairly happy with ticker handler, but this feels like a "six of one, half dozen of the other" type of scenario. It's not a one-off, and it's not going to go forever, but it probably IS going to go for quite a while, in some cases. I am currently considering completely disambiguating "active attack" wait state (e.g. fireball, trip, etc.) from the sort-of "passive attack" rounds with your weapon. I decided trying to link the two up was needlessly privileging adherence to Envy form over substance. Probably I'm going to use ticker handler for passive attacks and delay for waitstate (cooldown).
4. For the main thing I'm planning on using this for right now (movement), where it's fine if everyone makes the decision at the same time, probably I'll go tickerhandler.
5. This is almost exactly how I'd planned to do passive healing. :)) There will be in-combat healing, but quite slow. If you are full and sleeping, you'll heal fastest, down to starving and in-combat, as slowest.

**Comment by Griatch** [comment]
Great answers already in this thread! I want to point out that for `develop` branch, there is also a `utils.repeat` function that acts as a convenience wrapper for the `TickerHandler` in the same way that `utils.delay` acts as a convenience wrapper for the `TaskHandler`.

**Comment by Griatch** [comment]
> and was a bit confused by the "when not to use TickerHandler" at the bottom. I have several features that I am either in the middle of implementing, or am just about to implement, all of which involve some degree of timing.

I wrote that doc page so I can answer more specifically - the question is not only if your system involves some degree of timing, but if that timing should happen _without any input from the user_. The first and most efficient solution to consider (before deciding on ticker, delay or script described earlier in this thread) is the _on-demand_ solution. 

- Resetting areas: Technically, the area doesn't really need resetting unless there is a player there. So theoretically, its state (and time since last reset) could be checked when a new player enters and reset then. But that said, this could be an expensive operation so having it done async may still be worthwhile with a ticker - just make sure to not update the entire area every tick regardless of if things changed in it or not.
- Affects wearing off - this depends only on if you want to have a text `"<affect> wears off"` popping up or not. If you do, you need at least a delay-call. If not (such texts can get very spammy), you can just do it on-demand - whenever the player (or a rule check) needs the affected value, check the start time of the affect and compare it to the current time - disable it if the time passed is longer than the duration of the affect. Cheap and only requires storing the start time and a simple subtraction.
- Combat iteration - if you want mobs and players to perform visible actions on a repeating timer then yes, this sounds like a ticker is necessary. 
- Mobiles making decisions - this does require a ticker/script but it's also worth remembering the 'tree falling in the wood' thing - if there are no players around, do you really need to spend CPU cycles moving a mob around? While it may give you warm fuzzy feeling to have mobs move around 'in the world' even if noone else is around, this is still wasted resources if no players are around to see it.
- Healing/Mana/Recovery/Hunger/Thirst - again, the only thing that defines what to use here is if you want a text like "Your mana is fully recovered" to pop up without any user input. All other cases you can do on the fly: Set a healing rate and start-time, then when the level of health is relevant (either by you checking it or a system needing to know it), figure out how much time passed and multiply with the rate to find out your current health. How to deal with hunger/thirst is a special case - in reality you'll eventually starve to death by just standing around and doing nothing. But everyone hates the *you are feeling hungry* spam. So in-game you could also tie this to things that are actually affected by your hunger - such as moving or fighting - check the time since you last ate to figure out the penalty, if any - then report something like `"you move east (weakened by hunger)"` - the details obviously depend on your game.

**Reply by RandomEC** [reply]
1. Resetting areas. Yes, things only happen as far as resets if they need to. Everything checks to make sure it is where it needs to be/has what it needs, and if so, nothing happens on a reset. Resets happen, at the fastest, every five minutes.
2. Affects wearing off. I am planning on doing sort of a hybrid approach on this. Delay for removal of the actual affect (due to desire for a message), but players should be able to see how much time remains on it. This is going to be accomplished by storing the time on the affect, and when players check, it will just do the math.
3. Combat iteration - Yep.
4. Mobiles making decisions - You make a very good point here, and one that I've certainly thought a bit about. The issue is that if you don't let the mobiles move when people aren't around, then functionally, it looks like the reset condition every time you get there. So what I've been thinking about in the alternative is a) having non-sentinel mobiles only move when players are around, but b) having them start in semi-random rooms. Doing this gives the appearance that movement has been happening, but at a far lower cost.
5. Healing/Recovery - the issue here is that healing rate is dependent upon hunger/thirst, so I am a little worried about retroactively calculating it. Also, I'm going to run with a prompt, so every time the prompt comes up, it will need to be calculated. I guess the question is whether the convenience of just having the ticker handler do it is worth the trade-off. I am not worried about sending messages about hunger and thirst - if they want to know what it is, they can check it, and I agree those messages are annoying.
